File: server/src/main/java/com/chatroom/server/ChatServer.java
package com.chatroom.server;

import com.chatroom.shared.MessageProtocol;

import java.io.*;
import java.net.*;
import java.util.*;
import java.util.concurrent.*;

public class ChatServer {
    private static final int MAX_CLIENTS = 100;

    private final int port;
    private final ServerDisplay display;
    private final ConcurrentHashMap<String, ObjectOutputStream> clients = new ConcurrentHashMap<>();
    private final ConcurrentHashMap<Socket, ObjectOutputStream> outputStreams = new ConcurrentHashMap<>();
    private final ExecutorService pool = Executors.newFixedThreadPool(MAX_CLIENTS);
    private volatile ServerSocket serverSocket;
    private volatile boolean running = false;

    public ChatServer(int port, ServerDisplay display) {
        this.port = port;
        this.display = display;
    }

    public int getPort() {
        return serverSocket != null ? serverSocket.getLocalPort() : -1;
    }

    public void start() {
        try {
            serverSocket = new ServerSocket(port);
            running = true;
            display.showMessage("Waiting for clients at " + serverSocket + "\n");

            while (running) {
                Socket socket = serverSocket.accept();
                if (socket == null) break;
                try {
                    pool.execute(new ClientHandler(this, socket));
                } catch (RejectedExecutionException e) {
                    try { socket.close(); } catch (IOException ignored) {}
                    display.showMessage("Rejected connection (server full): " + socket + "\n");
                }
            }
        } catch (IOException e) {
            if (running) {
                display.showMessage("Server stopped or error: " + e.getMessage() + "\n");
            }
        }
    }

    void stopServer() {
        running = false;
        try {
            if (serverSocket != null && !serverSocket.isClosed()) {
                serverSocket.close();
            }
        } catch (IOException e) {
            // ignore
        }
        pool.shutdownNow();
    }

    void handleClientMessage(String username, Socket socket, String message) {
        if (MessageProtocol.isBroadcast(message)) {
            sendToAll(message);
        } else {
            String target = MessageProtocol.extractPrivateTarget(message);
            if (target != null) {
                int colon = message.indexOf(':');
                if (colon < 0) return;
                String content = message.substring(colon);
                String formattedMsg = "@" + username + content;
                sendPrivately(target, formattedMsg);
            }
        }
    }

    /** @return true if registered, false if username already taken */
    boolean registerClient(String username, Socket socket, ObjectOutputStream output) {
        if (clients.putIfAbsent(username, output) != null) {
            return false;
        }
        outputStreams.put(socket, output);
        display.showMessage("\n" + username + " (" + socket.getInetAddress().getHostAddress() + ") is online");
        broadcastUserList();
        return true;
    }

    void removeClient(String username) {
        clients.remove(username);
        broadcastUserList();
    }

    void removeConnection(Socket socket, String username) {
        outputStreams.remove(socket);
        display.showMessage("\n" + username + " (" + socket.getInetAddress().getHostAddress() + ") is offline");
    }

    void sendToAll(Object data) {
        // snapshot to avoid concurrent modification issues + weakly-consistent iteration
        var snapshot = new ArrayList<>(outputStreams.entrySet());
        for (var entry : snapshot) {
            ObjectOutputStream out = entry.getValue();
            synchronized (out) {
                try {
                    out.writeObject(data);
                    out.flush();
                } catch (IOException e) {
                    // evict broken client — will be cleaned up by ClientHandler finally block,
                    // but remove mapping now so user list stays accurate
                    outputStreams.remove(entry.getKey());
                }
            }
        }
    }

    void sendPrivately(String username, String message) {
        ObjectOutputStream output = clients.get(username);
        if (output != null) {
            synchronized (output) {
                try {
                    output.writeObject(message);
                    output.flush();
                } catch (IOException e) {
                    // client disconnected
                }
            }
        }
    }

    private void broadcastUserList() {
        sendToAll(MessageProtocol.buildUserList(new ArrayList<>(clients.keySet())));
    }
}

File: server/src/main/kotlin/com/chatroom/server/ChatServer.kt
package com.chatroom.server

import java.io.ObjectInputStream
import java.io.ObjectOutputStream
import java.net.ServerSocket
import java.net.Socket
import java.util.concurrent.ConcurrentHashMap
import kotlin.concurrent.thread

class ChatServer(private val port: Int, private val display: ServerDisplay) {
    private val clients = ConcurrentHashMap<String, ObjectOutputStream>()
    private val outputStreams = ConcurrentHashMap<Socket, ObjectOutputStream>()
    private var serverSocket: ServerSocket? = null
    @Volatile private var running = false

    fun getPort(): Int = serverSocket?.localPort ?: -1

    fun start() {
        try {
            serverSocket = ServerSocket(port)
            running = true
            display.showMessage("Waiting for clients at $serverSocket\n")

            while (running) {
                val socket = serverSocket?.accept() ?: break
                handleClient(socket)
            }
        } catch (e: Exception) {
            if (running) {
                display.showMessage("Server stopped or error: ${e.message}\n")
            }
        }
    }

    private fun handleClient(socket: Socket) {
        thread {
            var username: String? = null
            var output: ObjectOutputStream? = null
            try {
                // Important: Create ObjectOutputStream first and flush to write header
                output = ObjectOutputStream(socket.getOutputStream())
                output.flush()
                val input = ObjectInputStream(socket.getInputStream())

                // Read username
                val received = input.readObject()
                username = received.toString()

                // Register client
                clients[username] = output
                outputStreams[socket] = output

                // Broadcast user list
                sendToAll("!${clients.keys}")
                display.showMessage("\n$username (${socket.inetAddress.hostAddress}) is online")

                while (true) {
                    val message = input.readObject().toString()

                    if (message.contains("@EE@")) {
                        sendToAll(message)
                    } else {
                        // Private message format: @target: message
                        // We want to send to target: @sender: message
                        val splitIndex = message.indexOf(':')
                        if (splitIndex != -1) {
                            val target = message.substring(1, splitIndex)
                            val content = message.substring(splitIndex) // includes ": "
                            val formattedMsg = "@$username$content"
                            sendPrivately(target, formattedMsg)
                        }
                    }
                }

            } catch (e: Exception) {
                // Connection lost or closed
            } finally {
                username?.let {
                    removeClient(it)
                    removeConnection(socket, it)
                }
                try { socket.close() } catch (ignore: Exception) {}
            }
        }
    }

    private fun sendToAll(data: Any) {
        outputStreams.values.forEach { output ->
            synchronized(output) {
                try {
                    output.writeObject(data)
                    output.flush()
                } catch (e: Exception) {
                    // Handle broken pipe?
                }
            }
        }
    }

    private fun sendPrivately(username: String, message: String) {
        val output = clients[username]
        if (output != null) {
            synchronized(output) {
                try {
                    output.writeObject(message)
                    output.flush()
                } catch (e: Exception) {
                }
            }
        }
    }

    private fun removeClient(username: String) {
        clients.remove(username)
        sendToAll("!${clients.keys}")
    }

    private fun removeConnection(socket: Socket, username: String) {
        outputStreams.remove(socket)
        display.showMessage("\n$username (${socket.inetAddress.hostAddress}) is offline")
    }
}
